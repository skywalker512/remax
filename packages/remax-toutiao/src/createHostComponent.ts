import * as React from 'react';

export default function createHostComponent<P = any>(name: string) {
  const Component: React.FC<P> = (props, ref: any) => {
    const { children = [] } = props;
    return React.createElement(name, { ...props, ref }, children);
  };
  return React.forwardRef<{}, React.PropsWithChildren<P>>(Component);
}
