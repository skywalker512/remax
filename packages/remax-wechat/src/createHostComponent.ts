import * as React from 'react';

export default function createHostComponent<P = any>(
  name: string,
  component?: React.ComponentType<P>
) {
  if (component) {
    return component;
  }

  const Component: React.FC<P> = (props, ref: any) => {
    const { children = [] } = props;
    return React.createElement(name, { ...props, ref }, children);
  };
  return React.forwardRef<{}, React.PropsWithChildren<P>>(Component);
}
