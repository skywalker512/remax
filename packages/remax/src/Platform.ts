type Platform = 'alipay' | 'wechat' | 'toutiao';

const is = (platform: Platform) => process.env.REMAX_PLATFORM === platform;

const Platform = {
  get current() {
    return process.env.REMAX_PLATFORM;
  },
  get isAlipay() {
    return is('alipay');
  },
  get isWechat() {
    return is('wechat');
  },
  get isToutiao() {
    return is('toutiao');
  },
};

export default Platform;
