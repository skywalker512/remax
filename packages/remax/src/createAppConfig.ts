import * as React from 'react';
import render from './render';
import AppContainer from './AppContainer';
import isClass from './utils/isClass';
import isClassComponent from './utils/isClassComponent';
import { ForwardRef } from './ReactIs';

class DefaultAppComponent extends React.Component {
  render() {
    return React.createElement(React.Fragment, null, this.props.children);
  }
}

enum AppLifecyle {
  onLaunch = 'onLaunch',
  onShow = 'onShow',
  onHide = 'onHide',
  onError = 'onError',
  onShareAppMessage = 'onShareAppMessage',
  onPageNotFound = 'onPageNotFound',
}

export default function createAppConfig(this: any, App: any) {
  const createConfig = (
    AppComponent: React.ComponentType<any> = DefaultAppComponent
  ) => {
    const config = {
      _container: new AppContainer(this),

      _pages: [] as any[],

      _instance: React.createRef<any>(),

      onLaunch(options: any) {
        this._render();

        this.callLifecycle(AppLifecyle.onLaunch, options);
      },

      onShow(options: any) {
        this.callLifecycle(AppLifecyle.onShow, options);
      },

      onHide() {
        this.callLifecycle(AppLifecyle.onHide);
      },

      onError(error: any) {
        this.callLifecycle(AppLifecyle.onError, error);
      },

      // 支付宝
      onShareAppMessage(options: any) {
        this.callLifecycle(AppLifecyle.onShareAppMessage, options);
      },

      // 微信
      onPageNotFound(options: any) {
        this.callLifecycle(AppLifecyle.onPageNotFound, options);
      },

      callLifecycle(lifecycle: AppLifecyle, ...args: any[]) {
        if (this._instance.current && this._instance.current[lifecycle]) {
          return this._instance.current[lifecycle](...args);
        }
      },

      _mount(pageInstance: any) {
        this._pages.push(pageInstance);
        this._render();
      },

      _unmount(pageInstance: any) {
        this._pages.splice(this._pages.indexOf(pageInstance), 1);
        this._render();
      },

      _render() {
        const props: any = {};

        if (
          isClassComponent(AppComponent) ||
          (AppComponent as any).$$typeof === ForwardRef
        ) {
          props.ref = this._instance;
        }

        return render(
          React.createElement(
            AppComponent,
            props,
            this._pages.map(p => p.element)
          ),
          this._container
        );
      },
    };

    return config;
  };

  // 兼容老的写法
  if (isClass(App) && !isClassComponent(App)) {
    // eslint-disable-next-line no-console
    console.warn(
      '使用非 React 组件定义 App 的方式已经废弃，详细请参考：https://remaxjs.org/guide/framework'
    );
    Object.assign(App.prototype, createConfig());
    return new App();
  } else {
    return createConfig(App);
  }
}
