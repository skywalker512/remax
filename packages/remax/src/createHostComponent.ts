import * as React from 'react';

export const hostComponents: {
  [key: string]: {
    alias?: {
      [key: string]: string;
    };
  };
} = (process.env.REMAX_HOST_COMPONENTS as any) || {};

export default function createHostComponent<P = any>(
  name: string,
  component?: React.ComponentType<P>
) {
  if (component) {
    return component;
  }

  const Component: React.FC<P> = (props, ref: any) => {
    const { children = [] } = props;
    return React.createElement(name, { ...props, ref }, children);
  };
  return React.forwardRef<{}, React.PropsWithChildren<P>>(Component);
}
