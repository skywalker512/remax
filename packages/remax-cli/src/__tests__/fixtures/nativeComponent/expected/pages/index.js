'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('../npm/remax/esm/render.js');
var React = require('react');
require('../npm/remax/esm/createAppConfig.js');
require('../npm/remax/esm/createHostComponent.js');
require('../npm/remax/esm/Platform.js');
var createPageConfig = require('../npm/remax/esm/createPageConfig.js');
var createNativeComponent = require('../npm/remax/esm/createNativeComponent.js');
require('../npm/remax/esm/index.js');
var index$2 = require('../npm/remax/npm/remax-alipay/esm/hostComponents/View/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/ScrollView/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/Swiper/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/SwiperItem/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/MovableView/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/MovableArea/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/CoverView/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/CoverImage/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/Icon/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/Text/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/RichText/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/Progress/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/Button/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/CheckboxGroup/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/Checkbox/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/Form/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/Input/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/Label/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/Picker/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/PickerView/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/PickerViewColumn/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/RadioGroup/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/Radio/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/Slider/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/Switch/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/Textarea/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/Navigator/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/Image/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/Map/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/Canvas/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/WebView/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/Lifestyle/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/ContactButton/index.js');
require('../npm/remax/npm/remax-alipay/esm/hostComponents/Video/index.js');
require('../npm/remax/npm/remax-alipay/esm/api/index.js');

var CJSComponent = createNativeComponent.default('cjs-0');
var NotInJSXComponent = createNativeComponent.default('not-in-jsx-0');
var SrcComponent = createNativeComponent.default('src-0');
var ScopedComponent = createNativeComponent.default('a-1');
var SlotComponent = createNativeComponent.default('slot-0');
var PluginComponent2 = createNativeComponent.default('my-plugin-1');
var PluginComponent = createNativeComponent.default('my-plugin-0');
var Complex = createNativeComponent.default('complex-0');
var E = createNativeComponent.default('e-0');
var D = createNativeComponent.default('d-0');
var C = createNativeComponent.default('c-0');
var B = createNativeComponent.default('b-0');
var A = createNativeComponent.default('a-0');














var _page = function _page() {
  var b = React.createRef();
  var text = 'not in jsx' + NotInJSXComponent;
  return React.createElement(index$2.default, null, React.createElement(A, {
    foo: "bar"
  }), React.createElement(B, {
    ref: b
  }), React.createElement(C, null), React.createElement(D, null), React.createElement(E, null), React.createElement(Complex, null), React.createElement(PluginComponent, null), React.createElement(PluginComponent2, null), React.createElement(SlotComponent, null, React.createElement(index$2.default, {
    slot: "inner"
  })), React.createElement(ScopedComponent, null), React.createElement(SrcComponent, null), React.createElement(CJSComponent, null), text);
};

var index = Page(createPageConfig.default(_page));

exports.default = index;
