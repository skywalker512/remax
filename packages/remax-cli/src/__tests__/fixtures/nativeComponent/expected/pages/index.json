{
  "usingComponents": {
    "a-0": "../components/a/index",
    "a-1": "../components/_foo/a/index",
    "b-0": "../components/b/index",
    "c-0": "../components/c/index",
    "cjs-0": "../components/CJS/index",
    "complex-0": "../components/complex/index",
    "d-0": "../components/d/index",
    "e-0": "../components/e/index",
    "my-plugin-0": "plugin://myPlugin/xx",
    "my-plugin-1": "plugin://myPlugin/dddd",
    "not-in-jsx-0": "../components/notInJSX/index",
    "slot-0": "../components/slot/index",
    "src-0": "../components/src/index"
  }
}